<?xml version="1.0" encoding="utf-8"?>
<launch>
  ###########  launch config  ###########
  <arg name="real_machine" default="True" />
  <arg name="simulation" default="False" />
  <arg name="estimate_mode"  default= "0" />
  <arg name="sim_estimate_mode"  default= "2" />
  <arg name="headless" default="True" />
  <arg name="worldtype" default="$(find aerial_robot_simulation)/gazebo_model/world/empty.world" />
  <arg name="tree" default="real_RP_271_hokuyo_v1.1.3" />
  <arg name="tree_movement" default="dynamic" />
  <arg name="policy_start_delay" default="1.0" />
  <arg name="nav_exec" default="True" />
  <arg name="launch_gazebo" default="True" />
  <arg name="direct_model" default="False" />
  <arg name="direct_model_name" />
  <arg name="spawn_x" default="0.0"/>
  <arg name="spawn_y" default="0.0"/>
  <arg name="spawn_z" default="0.0"/>
  <!-- shift of obstacle position -->
  <arg name="spawn_yaw" default="0.0"/>
  <arg name="robot_id" default="" />
  <arg name="airframe" default="quadrotor" />
  <arg name="robot_ns" value="multirotor$(arg robot_id)" />
  <arg name="config_dir" default="$(find agile_multirotor)/config/" />
  <arg name="lidar" default="False" />
  <arg name="rate" default="20.0" />

  ###########  Parameters  ###########
  <group ns="$(arg robot_ns)">

    ###########  Basic Param  ###########
    # EGOMOTION_ESTIMATE = 0
    # EXPERIMENT_ESTIMATE = 1. for unstable mocap, use this mode
    # GROUND_TRUTH = 2
    <param name="estimation/mode" value= "$(arg sim_estimate_mode)" if="$(arg simulation)" />
    <param name="estimation/mode" value= "$(arg estimate_mode)" if="$(arg real_machine)"/>

    # Drone = 0
    <param name="uav_model" value= "0" />
    <rosparam file="$(arg config_dir)/RobotModelUnderActuated.yaml" command="load" />

    ###########  Motor Config  ###########
    <rosparam file="$(arg config_dir)/MotorInfo_F35.yaml" command="load" />

    ###########  Battery Config  ###########
    <rosparam file="$(arg config_dir)/Battery.yaml" command="load" />

    ###########  Control  ###########
    <rosparam file="$(arg config_dir)/FlightControlUnderActuated.yaml" command="load" />
    ###########  Sensor Fusion  ###########
    <rosparam file="$(arg config_dir)/StateEstimation.yaml" command="load" />
    <rosparam file="$(arg config_dir)/Simulation.yaml" command="load" if="$(eval arg('simulation') and not arg('real_machine'))"/>

    ###########  Navigation  ###########
    <rosparam file="$(arg config_dir)/NavigationConfig.yaml" command="load" />

    ###########  Obstacle Avoidance  ###########
    <rosparam file="$(arg config_dir)/ObstacleAvoidance.yaml" command="load" />
  </group>

  ###########  Base Platform  ###########
  <node pkg="aerial_robot_base" type="aerial_robot_base_node" name="aerial_robot_base_node" ns="$(arg robot_ns)" output="screen" >
    <param name="tf_prefix" value="$(arg robot_ns)"/>
    <param name="param_verbose" value="false"/>
    <param name="main_rate" value="40"/>
    <param name="nav_exec" value="$(arg nav_exec)"/>
  </node>

  ###########  Robot Model  ###########
  <arg name="description_mode" value="urdf" unless="$(arg simulation)" />
  <arg name="description_mode" value="gazebo" if="$(arg simulation)" />
  <include file="$(find aerial_robot_model)/launch/aerial_robot_model.launch" >
    <arg name="headless" value="$(arg headless)" />
    <arg name="robot_model" value="$(find agile_multirotor)/urdf/$(arg airframe).$(arg description_mode).xacro"/>
    <arg name="robot_model" value="$(arg direct_model_name)" if="$(arg direct_model)"/>
    <arg name="robot_ns" value="$(arg robot_ns)"/>
    <arg name="rviz_config" value="$(find agile_multirotor)/config/rviz_config" />
    <arg name="rviz_init_pose" value="$(arg config_dir)/RvizInit.yaml" />
    <arg name="need_joint_state" value="false" if ="$(eval arg('simulation') or arg('real_machine'))"/>
  </include>

  ###########  Sensors ###########
  <include file="$(find agile_multirotor)/launch/includes/sensors.launch.xml" >
    <arg name="real_machine" value="$(arg real_machine)" />
    <arg name="simulation" value="$(arg simulation)" />
    <arg name="robot_ns" value="$(arg robot_ns)" />
  </include >

  ########## Simulation in Gazebo #########
  <include file="$(find aerial_robot_simulation)/launch/gazebo.launch" if = "$(eval arg('simulation') and not arg('real_machine'))" >
    <arg name="robot_ns" default="$(arg robot_ns)" />
    <arg name="gui" default="false" if="$(arg headless)" />
    <arg name="headless" default="$(arg headless)" />
    <arg name="launch_gazebo" default="$(arg launch_gazebo)" />
    <arg name="worldtype" value="$(arg worldtype)" />
    <arg name="spawn_x" value="$(arg spawn_x)" />
    <arg name="spawn_y" value="$(arg spawn_y)" />
    <arg name="spawn_z" value="$(arg spawn_z)" />
    <arg name="spawn_yaw" value="$(arg spawn_yaw)" />
  </include>

  ######## work around to evoke robot model update #######
  <node pkg="agile_multirotor" type="joint_state_pub.py" name="joint_state_pub" ns="$(arg robot_ns)"/>

  ####### Simulation collision world ############
  <group if= "$(eval arg('simulation') and not arg('real_machine'))">
    <node pkg="agile_multirotor" type="obstacle_world.py" name="obstacle_world" output="screen" ns="$(arg robot_ns)">
      <param name="rate" value="$(arg rate)" />
      <param name="obstacle_world_file" value="$(find agile_multirotor)/config/obstacle_world/$(arg tree_movement)/multirotor/$(arg tree).csv" />
      <param name="policy_start_delay" value="$(arg policy_start_delay)" />
      <!-- <remap from="start_moving_obstacle" to="start_navigation" /> -->
    </node>

    <node pkg="agile_multirotor" type="depth_image_correction.py" name="depth_image_correction" output="screen" ns="$(arg robot_ns)"/> 
  </group>

  ####### Polar Pixel calculation  ############

  <!-- <node pkg="agile_multirotor" type="common_observation_conversion_node" name="common_observation_conversion" output="screen" ns="$(arg robot_ns)"> -->
  <!--   <param name="cfg_path" value="$(find agile_multirotor)/config/obstacle_world/$(arg tree_movement)/multirotor/$(arg tree).csv" /> -->
  <!--   <param name="hokuyo" type="bool" value="$(arg lidar)"/> -->
  <!--   <param name="shift_x" type="double" value="$(arg shift_x)" /> -->
  <!--   <param name="shift_y" type="double" value="$(arg shift_y)" /> -->
  <!--   <param name="robot_ns" value="$(arg robot_ns)"/> -->
  <!--   <param name="print_yaw" type="bool" value="false" /> -->
  <!--   <param name="vel_calc_boundary" type="double" value="0.30" /> -->
  <!--   <param name="body_r" type="double" value="$(arg body_r)" /> -->
  <!--   <param name="policy_start_delay" value="$(arg policy_start_delay)" /> -->
  <!--   <param name="headless" value="$(arg headless)" /> -->
  <!-- </node> -->
  <node pkg="agile_multirotor" type="raw_lidar_to_observation_node" name="raw_lidar_to_observation_node" output="screen" ns="$(arg robot_ns)">
    <param name="cfg_path" value="$(find agile_multirotor)/config/obstacle_world/$(arg tree_movement)/multirotor/$(arg tree).csv" />
    <param name="hokuyo" type="bool" value="$(arg lidar)"/>
    <param name="robot_ns" value="$(arg robot_ns)"/>
    <param name="print_yaw" type="bool" value="false" />
    <param name="vel_calc_boundary" type="double" value="0.30" />
    <param name="policy_start_delay" value="$(arg policy_start_delay)" />
    <param name="headless" value="$(arg headless)" />
  </node>
  <!-- <node pkg="agile_multirotor" type="policy_navigator.py" name="policy_navigator" output="screen" ns="$(arg robot_ns)" launch-prefix="xterm -fn 10x20 -e python -m pdb">
    <param name="ppo_path" value="$(find agile_multirotor)/linvel_policy" />
  </node> -->

  ####### HOKUYO sensor launch  ############
  <!-- real exec to use hokuyo sensor  -->
  <include file="$(find hokuyo_estimation)/launch/bringup.launch" if= "$(arg lidar)">
    <arg name="real_machine" value="$(arg real_machine)" />
    <arg name="simulation" value="$(arg simulation)" />
    <arg name="robot_ns" value="$(arg robot_ns)" />
    <arg name="sensor" value="livox"/>
  </include>

  ####### make laser_frame_level tf  ############
  <node pkg="agile_multirotor" type="laser_frame_level.py" name="laser_frame_level" output="screen" ns="$(arg robot_ns)"/>

</launch>
